community/kernel/src/main/java/org/neo4j/kernel/impl/proc/OutputMappers.java

```diff
@@ -0,0 +1,186 @@
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.neo4j.kernel.api.exceptions.ProcedureException;
+import org.neo4j.kernel.api.exceptions.Status;
+import org.neo4j.kernel.api.proc.ProcedureSignature;
+import org.neo4j.kernel.api.proc.ProcedureSignature.FieldSignature;
+
+import static java.lang.reflect.Modifier.isPublic;
+import static java.lang.reflect.Modifier.isStatic;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Takes user-defined record classes, and does two things: Describe the class as a {@link ProcedureSignature}, and provide a mechanism to convert
+ * an instance of the class to Neo4j-typed Object[].
+ */
+public class OutputMappers
+{
+    public OutputMappers( TypeMappers typeMappers )
+    {
+        this.typeMappers = typeMappers;
+    }
+
+    /**
+     * A compiled mapper, takes an instance of a java class, and converts it to an Object[] matching the specified {@link #signature()}.
+     */
+    public static class OutputMapper
+    {
+        private final List<FieldSignature> signature;
+        private final FieldMapper[] fieldMappers;
+
+        public OutputMapper( FieldSignature[] signature, FieldMapper[] fieldMappers )
+        {
+            this.signature = asList( signature );
+            this.fieldMappers = fieldMappers;
+        }
+
+        public List<FieldSignature> signature()
+        {
+            return signature;
+        }
+
+        public Object[] apply( Object record ) throws ProcedureException
+        {
+            Object[] output = new Object[fieldMappers.length];
+            for ( int i = 0; i < fieldMappers.length; i++ )
+            {
+                output[i] = fieldMappers[i].apply( record );
+            }
+            return output;
+        }
+    }
+
+    /**
+     * Extracts field value from an instance and converts it to a Neo4j typed value.
+     */
+    private static class FieldMapper
+    {
+        private final MethodHandle getter;
+        private final TypeMappers.NeoValueConverter mapper;
+
+        public FieldMapper( MethodHandle getter, TypeMappers.NeoValueConverter mapper )
+        {
+            this.getter = getter;
+            this.mapper = mapper;
+        }
+
+        Object apply( Object record ) throws ProcedureException
+        {
+            Object invoke = getValue( record );
+            return mapper.toNeoValue( invoke );
+        }
+
+        private Object getValue( Object record ) throws ProcedureException
+        {
+            try
+            {
+                return getter.invoke( record );
+            }
+            catch ( Throwable throwable )
+            {
+                throw new ProcedureException( Status.Procedure.CallFailed, throwable, "Unable to read value from record: %s", throwable.getMessage() );
```

Comment:
Looks like it can be useful to include record into exception in some way to be able to see what actually was received as record here